### PR TITLE
Update adding_an_op.md

### PR DESCRIPTION
**Pull request by ecpoppenheimer**
I am proposing several changes to this file.  I am a new user to TF, and used this documentation extensively while trying to put a together my custom activation function, since I wanted to explore how different activation functions might affect NN training and performace.  I eventually figured out everything I needed to know, but it took weeks because this documentation is lacking in several areas, so I have a good feel for how this documentation can be improved to make life easier for future new users.  Changes 1 and 2 are pretty clear and I expect to not be controversial.  Change 3 you will probably need to think about, and might want to do further work to make things more clear.

1) correct example.cc (221-228):
I added a REGISTER_OP macro to the "example" code.  The previous absence of these lines was clearly an omission in the documentaion - the example code does not work without this line.  This is the least controversial change I am making.

2) explain how to use Bazel to compile a CUDA op (1256-1272:
I eventually figured out how to do this on my own, but it sure would have been useful had this information been in this documentation.

3) explaining how the gradient actually works in TF (1293-1301):
Now that I know the answer, I can read the documentation and sort of parse what it means, but I found this extremely confusing when I was reading it for the first (and tenth) time.  I have a masters in physics, so I understand calculus, but it wasn't until I had implemented backpropagation twice by hand and spent a bunch of time trying to imagine how TF must be working under the hood before I got this.  So I was tempted to delete this entire section and start over.  But I am not sure I can win this battle, because a major portion of what makes this confusing for me is the way TF uses the word gradient.  The thing that all of the TF documentation refers to as "gradient" I would much prefer to refer to as "error".  But I doubt I am going to get everybody to change everything.

So I have opted to just append another paragraph to this section, which I hope will explain things better.  It might be nice to update the mathematical notation a little bit, but since I am currently making these edits in a simple text editor I don't want to wade too deeply into trying to figure out the notation.

4) Putting it all together (1467-1778):
I was able to find pieces of example code while googling, but somehow I wasn't really able to find a complete tutorial that documented all of the steps start to finish of designing and using a custom activation function.  So I think it might be useful to post a complete example here which is just complicated enough to demonstrate all the parts needed to write a custom activation function working together.  But you might decide that this belongs elsewhere.

### CODE REVIEW

**Comment by ecpoppenheimer** [comment]
I signed it!

File: tensorflow/docs_src/extend/adding_an_op.md [hunk 1]
```diff
@@ -1427,6 +1464,318 @@ compact in representing input and output shape specifications in tests.  For
 now, see the surrounding comments in those tests to get a sense of the shape
 string specification).
 
+## Putting it all together
+For the sake of ease and completeness, here is a full tutorial of how to implement a custom operator which is intended to be used as an activation function.  This isn't actually a very good activation function; it performs worse than commonly used activation functions like ReLu or Sigmoid, but this example is sufficient to use in a neural network and get it to train to marginal accuracy.  This specific mathematical form for the activation function was chosen only to demonstrate clearly how the gradient needs to be calculated, since ReLu is so simple that you can implement its gradient without really being clear what is happening from a mathematical standpoint
+
+1) Create the following three source files in the [`tensorflow/core/user_ops`][user_ops] directory:
+```c++
```
Comment (martinwicke) [inline tensorflow/docs_src/extend/adding_an_op.md]: Can you put code in code files and link to them in the md? I think that's better than adding them in the md. That'll just go out of date. 

Once you have done that, we absolutely need a test to make sure this actually works.

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
@MarkDaoust can you take a look? Some significant new tutorial content.

**Comment by ecpoppenheimer** [comment]
Sorry about the delay.  I hesitated when you first requested this change, since I didn't really know what to do, and kind of forgot this until you reminded me.

Code files:  I am not quite sure what you want me to do, since I am new to open source collaboration, but I my guess is that you want me to create three files with the smoothReLu kernel source code hosted somewhere on github and include links to those files in the md.  But if this is correct, where is the correct place to host the files?  Should they be stored somewhere within the TensorFlow project directory or should I host them on my personal github account.

Testing:  I did some testing, though perhaps not enough, but I don't really know what standards you hold for how much testing is sufficient.  I did test the files to make sure they compile and load in python properly.  I believe I tested the activation function in an actual neural network, which isn't sufficient to prove that it is working completely as desired, but does demonstrate that it is at least most of the way there.  I don't believe I actually tested that the function returns the correct values, and I didn't try to test any edge cases / behavior with invalid inputs.

**Comment by drpngx** [comment]
@martinwicke could you elaborate on the requested changes? It looks like @ecpoppenheimer needs more guidance.

**Comment by MarkDaoust** [comment]
Hi @ecpoppenheimer

> where is the correct place to host the files?

This directory:

https://github.com/tensorflow/models/tree/master/samples/core

Make a new sub-directory called `extend/` and put the files there. (We're trying to keep examples out of the main repo going forward.)

Assign that PR to me after you create it.

> I did some testing, though perhaps not enough.

The main thing we'd want here is a small test script that runs it a couple of times, so we know if it breaks. Something like [this](https://github.com/tensorflow/models/blob/master/tutorials/image/cifar10/cifar10_input_test.py), [or this](https://github.com/tensorflow/models/blob/master/tutorials/rnn/ptb/reader_test.py).

I'm not sure about the best way to setup the test to compile it, we'll ask the infrastructure people, but `subprocess.check_call` might be enough, we'll figure that out in the other PR.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@ecpoppenheimer Any progress on this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @benoitsteiner: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.
